File: Zoo_Python/animal.py
from abc import abstractmethod
from urllib.parse import MAX_CACHE_SIZE
from organism import Organism


class Animal(Organism):

    MAX_CHILD_POS_RAND_TRIES = 15

    @abstractmethod
    def _getChild():
        pass

    # def __init__(self, world):
    #     super().__init__(world)

    # def __init__(self, world, x, y):
    #     super().__init__(world, x, y)

    # def __init__(self, world, strength, initiative, x, y, age):
    #     super().__init__(world, strength, initiative, x, y, age)

    def takeAction(self):
        self.move()
        
    def collision(self, attacker):
        if type(self) == type(attacker) :
            self.reproduce()
            return

        if attacker.getStrength() > self._strength:
            attacker.changePosition( self.x, self.y )
            # log attack
            self._world.deleteOrganism(self)

    def reproduce(self):
        timer = 0
        childPos = self.getRandomMovePosition(self.x, self.y)

        while self._world.getOrganism(childPos[0], childPos[1]) != None:
            if timer > self.MAX_CHILD_POS_RAND_TRIES:
                return

            childPos[0] = self.x
            childPos[1] = self.y
            childPos = self.getRandomMovePosition(childPos[0], childPos[1])
            timer += 1
        
        self._world.addOrganism(type(self)(self._world, childPos[0], childPos[1]))

    def move(self):
        newPos = self.getRandomMovePosition(self.x, self.y)
        organism = self._world.getOrganism(newPos[0], newPos[1])

        if organism != None:
            organism.collision(self)
        else:
            self.changePosition(newPos[0], newPos[1])

File: Zoo_Java/src/Zoo/Animal.java
package Zoo;
public abstract class Animal extends Organism {

    public static final int MaxChildPositionRandomisingTries = 15;

    protected abstract Organism getChild(int x, int y);

    public Animal(World world) {
        super(world);
    }

    public Animal(World world, int x, int y) {
        super(world, x, y);
    }

    public Animal(World world, int strength, int initiative, int x, int y, int age) {
        super(world, strength, initiative, x, y, age);
    }

    @Override
    public void takeAction() {
        move();
    }

    @Override
    public void collision(Organism attacker) {
        if (this.getClass() == attacker.getClass()) {
            reproduce();
            return;
        }

        if (attacker.getStrength() > strength) {
            attacker.changePosition(x, y);
            world.logAttack(attacker, this);
            world.deleteOrganism(this);
        }
    }

    public void reproduce() {
        int timer = 0;
        int[] childPos = new int[2];

        do {
            if (timer > MaxChildPositionRandomisingTries) {
                return;
            }
            childPos[0] = x;
            childPos[1] = y;
            childPos = getRandomMovePosition(childPos[0], childPos[1]);
            timer++;
        } while (world.getOrganism(childPos[0], childPos[1]) != null);

        world.addOrganism(getChild(childPos[0], childPos[1]));
    }

    public void move() {
        int[] newPos;

	    newPos = getRandomMovePosition( x, y );

	    Organism organism = world.getOrganism( newPos[0], newPos[1] );

        if (organism != null)
        {
            organism.collision( this );
        }
        else
        {
            changePosition( newPos[0], newPos[1] );
        }
    }
}
